driver/src/orderbook/streamed/state.rs

```diff
@@ -0,0 +1,561 @@
+        self.balances
+            .entry((event.user, event.token))
+            .or_default()
+            .deposit(
+                Flux {
+                    amount: event.amount,
+                    batch_id: event.batch_id,
+                },
+                block_batch_id,
+            );
+        Ok(())
+    }
+
+    fn withdraw_request(&mut self, event: &WithdrawRequest) -> Result<(), Error> {
+        let balance = self.balances.entry((event.user, event.token)).or_default();
+        balance.pending_withdraw = Some(Flux {
+            amount: event.amount,
+            batch_id: event.batch_id,
+        });
+        Ok(())
+    }
+
+    fn withdraw(&mut self, event: &Withdraw, block_batch_id: BatchId) -> Result<(), Error> {
+        let balance = self.balances.entry((event.user, event.token)).or_default();
+        balance.update_deposit_balance(block_batch_id);
+        balance.pending_withdraw = None;
+        balance.balance = balance
+            .balance
+            .checked_sub(event.amount)
+            .ok_or(Error::MathUnderflow)?;
+        Ok(())
+    }
+
+    fn token_listing(&mut self, event: &TokenListing) -> Result<(), Error> {
+        self.tokens.0.push((event.id, event.token));
+        Ok(())
+    }
+
+    fn order_placement(&mut self, event: &OrderPlacement) -> Result<(), Error> {
+        let order = Order {
+            buy_token: event.buy_token,
+            sell_token: event.sell_token,
+            valid_from: event.valid_from,
+            valid_until: event.valid_until,
+            price_numerator: event.price_numerator,
+            price_denominator: event.price_denominator,
+            used_amount: 0,
+        };
+        match self.orders.entry((event.owner, event.index)) {
+            Entry::Vacant(entry) => entry.insert(order),
+            Entry::Occupied(_) => return Err(Error::OrderAlreadyExists),
+        };
+        Ok(())
+    }
+
+    fn order_cancellation(
+        &mut self,
+        event: &OrderCancellation,
+        block_batch_id: BatchId,
+    ) -> Result<(), Error> {
+        let order = self
+            .orders
+            .get_mut(&(event.owner, event.id))
+            .ok_or(Error::UnknownOrder)?;
+        order.valid_until = block_batch_id - 1;
+        Ok(())
+    }
+
+    fn order_deletion(&mut self, event: &OrderDeletion) -> Result<(), Error> {
+        // Orders are allowed to be deleted multiple times.
+        self.orders.remove(&(event.owner, event.id));
+        Ok(())
+    }
+
+    fn solution_submission(
+        &mut self,
+        event: &SolutionSubmission,
+        block_batch_id: BatchId,
+    ) -> Result<(), Error> {
+        let fee_token = self.tokens.address(0).ok_or(Error::UnknownToken)?;
+        let balance = self
+            .balances
+            .entry((event.submitter, fee_token))
+            .or_default();
+        balance.update_deposit_balance(block_batch_id);
+        balance.balance += event.burnt_fees;
```

Comment:
Similar to https://github.com/gnosis/dex-services/pull/679#discussion_r403889834, here as well we should count this as a "deposit" since the fees are not yet spendable in the current batch.

Reply:
I think I have a good solution for these pending balances from solution submission and trades. Since that is related to how trades are handled I will move this in the other PR and this PR will have solution submission unimplemented.